firewood/src/merkle/stream.rs

```diff
@@ -396,6 +396,140 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
     }
 }
 
+enum PathIteratorState<'a> {
+    Iterating {
+        /// The key, as nibbles, of the node at [address], without the
+        /// node's partial path (if any) at the end.
+        /// Invariant: If this node has a parent, the parent's key is a
+        /// prefix of the key we're seeking.
+        /// Note the node at [address] may not have a key which is a
+        /// prefix of the key we're seeking.
+        matched_key: Vec<u8>,
+        unmatched_key: NibblesIterator<'a, 0>,
+        address: DiskAddress,
+    },
+    Exhausted,
+}
+
+/// Iterates over all nodes on the path to a given key starting from the root.
+/// If the key isn't in the trie, the final returned node is either the node with the
+/// longest prefix of the key or the node that is where the key would be inserted if
+/// it were in the trie. Note this means the final returned node may not be a prefix
+/// of the key we're seeking.
```

Comment:
I think I understand the point above now. If my understanding is correct, the only time the returned key *is not* a prefix is when it's the initial key the iterator was created with.

If that's the case, it would be better to say the key for each node could be one of two things, either the key we're iterating over or a prefix of that key. 

However, I might be missing something here...